yank.1
.Dd $Mdocdate: July 10 2015 $
.Dt YANK 1
.Os
.Sh NAME
.Nm yank
.Nd yank terminal output to clipboard
.Sh SYNOPSIS
.Nm
.Op Fl 1ilxv
.Op Fl d Ar delim
.Op Fl g Ar pattern
.Op Fl - Ar command Op Ar argument ...
.Sh DESCRIPTION
Read input from
.Pa stdin
and display a selection interface that allows a field to be selected and copied
to the clipboard.
Fields are either recognized by a regular expression using the
.Fl g
option or by splitting the input on a delimiter sequence using the
.Fl d
option, see
.Sx DELIMITERS .
.Pp
Using the arrow keys will move the selected field, see
.Sx COMMANDS .
Pressing the return key will invoke
.Ar command
and write the selected field to its
.Pa stdin .
The
.Ar command
defaults to
.Xr xsel 1x
but could be anything that accepts input on
.Pa stdin ,
see
.Sx EXAMPLES .
.Pp
The options are as follows:
.Bl -tag -width Ds
.It Fl 1
If only one field is recognized, select it and invoke
.Ar command
without displaying the selection interface.
.It Fl d Ar delim
All input characters not present in
.Ar delim
will be recognized as fields, see
.Sx DELIMITERS .
.It Fl g Ar pattern
Use
.Ar pattern
to recognize fields, expressed as a POSIX extended regular expression.
.It Fl i
Ignore case differences between
.Ar pattern
and the input.
.It Fl l
Use the default delimiters except for space, see
.Sx DELIMITERS .
.It Fl v
Prints version.
.It Fl x
Use alternate screen.
.It Fl - Ar command Op Ar argument ...
Use
.Ar command
with zero or more
.Ar args
as the yank command.
.El
.Sh COMMANDS
.Bl -tag -width XXXX
.It Ic Ctrl-A | g
Move selection to the first field.
.It Ic Ctrl-C | Ctrl-D
Exit without invoking the yank command.
.It Ic Ctrl-E | G
Move selection to the last field.
.It Ic Ctrl-P Ns / Ns Ic Ctrl-N | Ic Left Ns / Ns Ic Right | Ic h Ns / Ns Ic l
Move selection to the left or right.
.It Ic Up Ns / Ns Ic Down | Ic j Ns / Ns Ic k
Move selection to the next or previous line.
.It Ic Enter
Exit using the selected field.
.El
.Sh DELIMITERS
If the
.Fl d
and
.Fl g
options are omitted the following characters are recognized as delimiters by
default:
.Bl -tag -width XXXX
.It \ef
form feed
.It \en
new line
.It \er
carriage return
.It \es
space
.It \et
horizontal tab
.El
.Pp
If the
.Fl d
option is present space is not recognized as a delimiter.
.Sh EXAMPLES
Yank an environment variable key or value:
.Pp
.Dl $ env | yank -d =
.Pp
Yank a field from a CSV file:
.Pp
.Dl $ yank -d \e\&", <file.csv
.Pp
Yank a whole line using the
.Fl l
option:
.Pp
.Dl $ make 2>&1 | yank -l
.Pp
If
.Pa stdout
is not a terminal the selected field will be written to
.Pa stdout
and exit without invoking the yank command.
Kill the selected PID:
.Pp
.Dl $ ps ux | yank -g [0-9]+ | xargs kill
.Pp
Yank the selected field to the clipboard as opposed of the default primary
clipboard:
.Pp
.Dl $ yank \(em xsel -b
.Sh DIAGNOSTICS
.Ex -std
.Sh SEE ALSO
.Xr re_format 7
.Sh AUTHORS
.An Anton Lindqvist Aq Mt anton@basename.se
.Sh CAVEATS
Recognizing fields enclosed in brackets requires
.Sq \&]
to be present before
.Sq \&[
in the argument given to the
.Fl d
option,
see
.Xr re_format 7 .


CHANGELOG.md
# v1.3.0 - 2022-06-18

## News

- Add opt-in selection of one and only field.
  (d8b2024)
  (Anton Lindqvist)

# v1.2.0 - 2019-08-12

- Add openSUSE installation instructions to README.
  (c4ce72e)
  (Dany Marcoux)

- Do not confuse the escape key with up arrow.
  (8536116)
  (Emanuele Giaquinta)

- Remove opinionated `-Werror` option from `CFLAGS`.
  (f9ac9e6)
  (Anton Lindqvist)

- Map carriage return to newline.
  (dffe105)
  (Anton Lindqvist)

# v1.1.0 - 2018-11-01

- Fix compilation on FreeBSD by passing preprocessor macros in `CFLAGS` as
  opposed of using `CPPFLAGS` which is not honored.
  (78ededb)
  (Tobias Kortkamp)

- Add FreeBSD install instructions.
  (0dd944b)
  (Tobias Kortkamp)

- Fix a bogus free by storing a copy of the pattern argument.
  (8dff9f3)
  (Tobias Kortkamp)

# v1.0.0 - 2018-09-09

- Fix a broken reference and improve the semantics of the manual.
  (f99c6ed, 7a964c1)
  (Anton Lindqvist)

- Add Nix installation instructions to README.
  (412a5cd, #40)
  (Matt)

- Add caveats section to manual.
  (88e8ca8)
  (Anton Lindqvist)

- Start providing deterministic release tarballs,
  instead of relying on the ones created by GitHub.
  (99a91e2)

# v0.8.3 - 2017-06-05

- Fix typo in README and man page (fe45fc1) (Jakub Wilk)

- Add OpenBSD install instructions to README (c20e9bb) (Anton Lindqvist)

- Call `setlocale(3)` prior pledge, allows the `rpath` promise to be removed
  (20cdc5a) (Anton Lindqvist)

- Compile and link separately (7691802) (Anton Lindqvist)

- Cleanup README and man page (495b53e, 6dd722d) (Anton Lindqvist)

- `snprintf(3)` can return `-1` (9ecb1f5) (Anton Lindqvist)

# v0.8.2 - 2017-02-19

- Ensure `make clean` always exits with success (47a0ee9) (Sébastien Delafond)

# v0.8.1 - 2017-02-18

- Add Fedora install instructions to README (#33, #34, 8d90aae) (Nemanja
  Milosevic)

- Allow the name of the compiled binary to be changed using the `PROG` variable
  (#35, #36, 1c43c46) (Nemanja Milosevic)

# v0.8.0 - 2017-01-17

- Update man page and README (9eac46f, 252a419) (Anton Lindqvist)

- Add Arch Linux install instructions to README (#32, db259d8) (Javier Tiá)

- Use `install(1)` in Makefile (11a2dc1, 42604e9) (Anton Lindqvist)

- Add support for `DESTDIR` to Makefile (77cf987) (Anton Lindqvist)

# v0.7.1 - 2016-10-08

- Update man page (2e78c28, a5cdba9) (Anton Lindqvist)

- Consolidate key bindings (5b4a840) (Anton Lindqvist)

- Fix `snprintf(3)` return value check (9fbf248) (Anton Lindqvist)

- Improve error messages (89be934) (Anton Lindqvist)

- Fix segfault when no fields are found (f3ef1ea) (Anton Lindqvist)

- Add Vi goto `g` and `G` key bindings (859f27b) (Anton Lindqvist)

- Refactoring and cleanup (2017de6) (Anton Lindqvist)

# v0.7.0 - 2016-05-18

- Add MANPREFIX support to installation task (Anton Lindqvist)

- Pledge on OpenBSD (Anton Lindqvist)

# v0.6.4 - 2016-03-13

- Add Vi key bindings (Samson Yeung)

- Fix wrong exit code on yank command failure (Anton Lindqvist)

# v0.6.3 - 2016-02-20

- Exit with correct code on invocation failure (Anton Lindqvist)

- Don't allow yanking if the input doesn't contain any fields (Anton Lindqvist)

- Fix endless loop while executing the supplied regex (Anton Lindqvist)

# v0.6.2 - 2015-12-15

- Fix non string literal compilation warning (Anton Lindqvist)

# v0.6.1 - 2015-12-15

- Use more common save/restore cursor capabilities (Eli Young)

# v0.6.0 - 2015-12-06

- Add regex delimiter support using the `-g` option and `-i` to make the pattern
  case-insensitive (Anton Lindqvist)

# v0.5.0 - 2015-11-12

- Add arrow key bindings (Anton Lindqvist)

- Add Debian installation instructions (Sébastien Delafond)

- Add OS X installation instructions (Carl Dong)

# v0.4.1 - 2015-10-19

- Fix number of lines in input bug (Anton Lindqvist)

- Fix typo in man page (FND)

- Fix description in man page (Jeremiah LaRocco)

# v0.4.0 - 2015-10-10

- Always recognize carriage return as a delimiter (Anton Lindqvist)

- Exit with the same exit code as the yank command on failure (Anton Lindqvist)

- Ensure all data is written to the terminal (Anton Lindqvist)

- Fix number of lines in input bug (Anton Lindqvist)

# v0.3.1 - 2015-09-24

- Fix moving beyond input bug (Anton Lindqvist)

# v0.3.0 - 2015-09-21

- Add `-x` option used to enable usage of the alternate screen (Anton Lindqvist)

- Fix `style(9)` violations (Anton Lindqvist)

# v0.2.0 - 2015-09-10

- Add `-l` option used to yank a whole line (Anton Lindqvist)

- Add `Ctrl-D` key binding which will exit without invoking the yank command
  (Anton Lindqvist)

- Require presence of the `--` option used to separate options and the yank
  command (Anton Lindqvist)

- Fix number of lines in input bug (Anton Lindqvist)

# v0.1.1 - 2015-08-28

- Add support for linking against musl (Anton Lindqvist)

# v0.1.0 - 2015-08-24

- Initial commit (Anton Lindqvist)


yank.c
#include <sys/ioctl.h>
#include <sys/wait.h>

#include <err.h>
#include <errno.h>
#include <fcntl.h>
#include <getopt.h>
#include <locale.h>
#include <regex.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <termios.h>
#include <unistd.h>

/* Terminal capabilities */
#define T_CLR_EOS             "\033[J"
#define T_CURSOR_INVISIBLE    "\033[?25l"
#define T_CURSOR_VISIBLE      "\033[?25h"
#define T_ENTER_CA_MODE       "\033[?1049h"
#define T_ENTER_STANDOUT_MODE "\033[7m"
#define T_EXIT_CA_MODE        "\033[?1049l"
#define T_EXIT_STANDOUT_MODE  "\033[0m"
#define T_RESTORE_CURSOR      "\0338"
#define T_SAVE_CURSOR         "\0337"

#define MAX(x, y) ((x) > (y) ? (x) : (y))
#define MIN(x, y) ((x) < (y) ? (x) : (y))

enum {
	KEY_ENTER = 1,
	KEY_HOME,
	KEY_END,
	KEY_TERM,
	KEY_UP,
	KEY_RIGHT,
	KEY_DOWN,
	KEY_LEFT
};

struct field {
	size_t	so; /* start offset */
	size_t	eo; /* end offset */
	size_t	lo; /* line offset */
};

static regex_t reg;

static char **yankargv;

static struct {
	size_t		 nmemb;
	size_t		 size;
	struct field	*v;
} f;

static struct {
	size_t	 size;
	size_t	 nmemb;
	char	*v;
} in;

static struct {
	int		rfd;
	int		wfd;
	int		ca;	/* use alternate screen */
	struct termios	attr;
} tty;

static void
input(void)
{
	int n;

	in.size = BUFSIZ;
	if ((in.v = malloc(in.size)) == NULL)
		err(1, NULL);

	while ((n = read(0, in.v + in.nmemb, in.size - in.nmemb)) != 0) {
		if (n == -1)
			err(1, "read");
		in.nmemb += n;

		if (in.nmemb < in.size)
			continue;
		in.size *= 2;
		if ((in.v = realloc(in.v, in.size)) == NULL)
			err(1, NULL);
	}
	memset(in.v + in.nmemb, 0, in.size - in.nmemb);
}

/*
 * Returns s transformed into a negation regular expression pattern concatenated
 * with the default delimiters.
 */
static char *
strtopat(const char *s)
{
	const char *fmt = "[^%s\f\n\r\t]+";
	char *pat;
	size_t len;
	int n;

	len = strlen(s) + strlen(fmt) + 1;
	if ((pat = malloc(len)) == NULL)
		err(1, NULL);
	n = snprintf(pat, len, fmt, s);
	if (n == -1 || n >= (ssize_t)len)
		errx(1, "pattern too long");

	return pat;
}

/*
 * Returns zero if f1 and f2 intersects, 1 if f2 begins after f1 and -1
 * otherwise. Both field start and end offsets are normalized with their
 * corresponding line offset.
 */
static int
fcmp(const struct field *f1, const struct field *f2)
{
	size_t e1, e2, s1, s2;

	s1 = f1->so - f1->lo, e1 = f1->eo - f1->lo;
	s2 = f2->so - f2->lo, e2 = f2->eo - f2->lo;

	return MAX(s1, s2) <= MIN(e1, e2) ? 0 : (e1 < s2 ? 1 : -1);
}

static ssize_t
xwrite(int fd, const char *s, size_t nmemb)
{
	ssize_t r;
	size_t n;

	n = nmemb;
	do {
		r = write(fd, s, n);
		if (r == -1)
			return r;
		n -= r;
		s += r;
	} while (n);

	return nmemb;
}

static void
yank(const char *s, size_t nmemb)
{
	int fd[2];
	int status;
	pid_t pid;

	if (!isatty(1)) {
		if (xwrite(1, s, nmemb) == -1)
			err(1, "write");
		exit(0);
	}

	if (pipe(fd) == -1)
		err(1, "pipe");
	if (dup2(fd[0], 0) == -1)
		err(1, "dup2");
	if (close(fd[0]) == -1)
		err(1, "close");
	if (xwrite(fd[1], s, nmemb) == -1)
		err(1, "write");
	if (close(fd[1]) == -1)
		err(1, "close");
	pid = fork();
	switch (pid) {
	case -1:
		err(1, "fork");
		/* NOTREACHED */
	case 0:
		execvp(yankargv[0], yankargv);
		err(126 + (errno == ENOENT), "%s", yankargv[0]);
		/* NOTREACHED */
	default:
		if (waitpid(pid, &status, 0) == -1)
			err(1, "waitpid");
		if (WIFSIGNALED(status))
			exit(128 + WTERMSIG(status));
		if (WIFEXITED(status))
			exit(WEXITSTATUS(status));
	}
}

static void
twrite(const char *s, size_t nmemb)
{
	if (xwrite(tty.wfd, s, nmemb) == -1)
		err(1, "write");
}

static void
tputs(const char *s)
{
	size_t n;

	n = strlen(s);
	twrite(s, n);
}

static void
tsetup(void)
{
	struct termios attr;
	struct winsize ws;
	regmatch_t r;
	char *e, *s;
	size_t m, n, w;
	unsigned int i, j;

	if ((tty.rfd = open("/dev/tty", O_RDONLY)) == -1)
		err(1, "/dev/tty");
	if ((tty.wfd = open("/dev/tty", O_WRONLY)) == -1)
		err(1, "/dev/tty");

	if (ioctl(tty.rfd, TIOCGWINSZ, &ws) == -1)
		err(1, "TIOCGWINSZ");

	f.size = 32;
	if ((f.v = malloc(f.size*sizeof(struct field))) == NULL)
		err(1, NULL);
	m = n = MIN(ws.ws_col*ws.ws_row, (ssize_t)in.nmemb);
	s = e = in.v;
	while (m && !regexec(&reg, e, 1, &r, 0) && r.rm_eo - r.rm_so) {
		f.v[f.nmemb].so = f.v[f.nmemb].eo = e - s;
		f.v[f.nmemb].so += r.rm_so;
		f.v[f.nmemb].eo += MAX(MIN(r.rm_eo, (ssize_t)m) - 1, 0);
		e += r.rm_eo;
		m -= MIN(r.rm_eo, (ssize_t)m);

		if (++f.nmemb < f.size)
			continue;
		f.size *= 2;
		if ((f.v = realloc(f.v, f.size*sizeof(struct field))) == NULL)
			err(1, NULL);
	}

	for (i = j = 0, s = e = in.v; n && i < ws.ws_row; i++) {
		if (s == e && !(e = memchr(s + 1, '\n', n)))
			e = in.v + in.nmemb;

		w = MIN(e - s, ws.ws_col);
		for (; j < f.nmemb && f.v[j].so < (size_t)(s - in.v + w); j++)
			f.v[j].lo = s - in.v;
		s += w;
		n -= w;
	}
	f.nmemb = MIN(f.nmemb, j);
	/* Ensure last field does not exceed the terminal width. */
	if (n > 0 && f.nmemb > 0 &&
	    f.v[f.nmemb - 1].eo - f.v[f.nmemb - 1].lo >= ws.ws_col)
		f.v[f.nmemb - 1].eo = f.v[f.nmemb - 1].lo + ws.ws_col - 1;
	/* Number of bytes to output. */
	f.v[f.nmemb].lo = MAX(s - in.v - 1, 0);

	if (tcgetattr(tty.rfd, &tty.attr) == -1)
		err(1, "tcgetattr");
	attr = tty.attr;
	attr.c_iflag |= ICRNL;
	attr.c_lflag &= ~(ICANON|ECHO|ISIG);
	if (tcsetattr(tty.rfd, TCSANOW, &attr) == -1)
		err(1, "tcsetattr");

	if (tty.ca)
		tputs(T_ENTER_CA_MODE);
	tputs(T_CURSOR_INVISIBLE);
	/* Emit the number of lines and save the cursor position. */
	for (j = 0; j < i; j++)
		tputs("\n");
	for (j = 0; j < i; j++)
		tputs("\033M");
	tputs(T_SAVE_CURSOR);
}

static void
tend(void)
{
	tputs(T_RESTORE_CURSOR);
	tputs(T_CLR_EOS);
	tputs(T_CURSOR_VISIBLE);
	if (tty.ca)
		tputs(T_EXIT_CA_MODE);
	tcsetattr(tty.rfd, TCSANOW, &tty.attr);
	close(tty.rfd);
	close(tty.wfd);
}

static int
tgetc(void)
{
	static struct {
		const char	*s;
		int		 c;
	} keys[] = {
		{ "\n",		KEY_ENTER },
		{ "\001",	KEY_HOME },	/* Ctrl-A */
		{ "\003",	KEY_TERM },	/* Ctrl-C */
		{ "\004",	KEY_TERM },	/* Ctrl-D */
		{ "\005",	KEY_END },	/* Ctrl-E */
		{ "\016",	KEY_RIGHT },	/* Ctrl-N */
		{ "\020",	KEY_LEFT },	/* Ctrl-P */
		{ "G",		KEY_END },
		{ "g",		KEY_HOME },
		{ "h",		KEY_LEFT },
		{ "j",		KEY_DOWN },
		{ "k",		KEY_UP },
		{ "l",		KEY_RIGHT },
		{ "\033[A",	KEY_UP },
		{ "\033[C",	KEY_RIGHT },
		{ "\033[B",	KEY_DOWN },
		{ "\033[D",	KEY_LEFT },
		{ NULL,		0 },
	};
	char buf[4];
	ssize_t n;
	int i;

	n = read(tty.rfd, buf, sizeof(buf) - 1);
	if (n == -1)
		err(1, "read");
	if (n == 0)
		return KEY_TERM;	/* EOF */
	buf[n] = '\0';

	for (i = 0; keys[i].s != NULL; i++)
		if (strncmp(keys[i].s, buf, strlen(keys[i].s)) == 0)
			return keys[i].c;

	return 0;
}

static const struct field *
tmain(void)
{
	size_t n;
	int c, i, j, k;

	i = j = 0;
	n = f.v[f.nmemb].lo;
	for (;;) {
		tputs(T_RESTORE_CURSOR);
		if (f.nmemb > 0) {
			twrite(in.v, f.v[i].so);
			tputs(T_ENTER_STANDOUT_MODE);
			twrite(in.v + f.v[i].so, f.v[i].eo - f.v[i].so + 1);
			tputs(T_EXIT_STANDOUT_MODE);
			twrite(in.v + f.v[i].eo + 1, n - f.v[i].eo);
		} else {
			twrite(in.v, n);
		}

		c = tgetc();
		switch (c) {
		case KEY_ENTER:
			if (f.nmemb > 0)
				return &f.v[i];
			break;
		case KEY_TERM:
			return NULL;
		case KEY_HOME:
			j = 0;
			break;
		case KEY_RIGHT:
			j = i + 1;
			break;
		case KEY_END:
			j = f.nmemb - 1;
			break;
		case KEY_LEFT:
			j = i - 1;
			break;
		case KEY_DOWN:
		case KEY_UP:
			if (c == KEY_DOWN) {
				j = i;
				while (j < (ssize_t)f.nmemb &&
				    f.v[i].lo == f.v[j].lo)
					j++;
				if (j == (ssize_t)f.nmemb)
					break;
			} else {
				k = i;
				while (k && f.v[i].lo == f.v[k].lo)
					k--;
				j = k;
				while (j && f.v[j - 1].lo == f.v[k].lo)
					j--;
			}
			for (; fcmp(&f.v[i], &f.v[j]) < 0 &&
			    f.v[j].lo == f.v[j + 1].lo; j++)
				continue;
			break;
		}
		if (j >= 0 && j < (ssize_t)f.nmemb)
			i = j;
	}
}

static void
usage(void)
{
	fprintf(stderr, "usage: yank [-1ilxv] [-d delim] [-g pattern] "
	    "[-- command [args]]\n");
	exit(2);
}

int
main(int argc, char *argv[])
{
	const struct field *field;
	char *pat;
	int one = 0;
	int rflags = REG_EXTENDED;
	int c, i;

	setlocale(LC_CTYPE, "");

#ifdef __OpenBSD__
	if (pledge("stdio tty proc exec", NULL) == -1)
		err(1, "pledge");
#endif

	pat = strtopat(" ");
	while ((c = getopt(argc, argv, "1ilvxd:g:")) != -1)
		switch (c) {
		case '1':
			one = 1;
			break;
		case 'd':
			free(pat);
			pat = strtopat(optarg);
			break;
		case 'g':
			free(pat);
			if ((pat = strdup(optarg)) == NULL)
				err(1, NULL);
			rflags |= REG_NEWLINE;
			break;
		case 'i':
			rflags |= REG_ICASE;
			break;
		case 'l':
			free(pat);
			pat = strtopat("");
			break;
		case 'v':
			puts("yank " VERSION);
			exit(0);
		case 'x':
			tty.ca = 1;
			break;
		default:
			usage();
		}
	argc -= optind;
	argv += optind;

	if (regcomp(&reg, pat, rflags) != 0)
		errx(1, "invalid regular expression");

	/* Ensure space for yank command and null terminator. */
	if ((yankargv = calloc(argc + 2, sizeof(char *))) == NULL)
		err(1, NULL);
	yankargv[0] = YANKCMD;
	for (i = 0; i < argc; i++)
		yankargv[i] = argv[i];

	input();
	tsetup();
	if (one && f.nmemb == 1)
		field = &f.v[0];
	else
		field = tmain();
	tend();
	if (field == NULL)
		return 1;
	yank(in.v + field->so, field->eo - field->so + 1);

	return 0;
}
